### PR TITLE
install native and cross lua executables & add LuaRocks plugin

### PR DESCRIPTION
**Pull request by starius**

### CODE REVIEW

File: plugins/luarocks/luarocks.mk [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+define $(PKG)_BUILD_SHARED
+    cd '$(1)' && ./configure \
+        --prefix='$(PREFIX)/$(TARGET)' \
+        --rocks-tree='$(PREFIX)/$(TARGET)' \
+        --lua-version='$(lua_SHORTVER)' \
+        --with-lua='$(PREFIX)/$(TARGET)' \
+        --with-lua-bin='$(PREFIX)/bin' \
+        --with-downloader='wget' \
+        --with-md5-checker='openssl'
+    $(MAKE) -C '$(1)' build install \
```
Comment (tonytheodore) [inline plugins/luarocks/luarocks.mk]: `install` doesn't depend on `build` so (on OSX) the symlink is created first and doesn't have execute permissions. These steps should be separated.

File: plugins/luarocks/luarocks.mk [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+PKG             := luarocks
+$(PKG)_IGNORE   :=
+$(PKG)_VERSION  := 2.2.2
+$(PKG)_CHECKSUM := 4f0427706873f30d898aeb1dfb6001b8a3478e46a5249d015c061fe675a1f022
+$(PKG)_SUBDIR   := luarocks-$($(PKG)_VERSION)
+$(PKG)_FILE     := $($(PKG)_SUBDIR).tar.gz
+$(PKG)_URL      := https://keplerproject.github.io/luarocks/releases/$($(PKG)_FILE)
+$(PKG)_DEPS     := gcc lua
+
+lua_SHORTVER    := $(call SHORT_PKG_VERSION,lua)
```
Comment (tonytheodore) [inline plugins/luarocks/luarocks.mk]: Move to `lua.mk`

**Comment by tonytheodore** [comment]
Looks good!

Plugins currently require some boilerplate:
```
$(PKG)_MAKEFILE := $(lastword $(MAKEFILE_LIST))
```
for updates to work, I might see if there's some sort of `include` hook so it isn't necessary.

Additional metadata like:
```
$(PKG)_WEBSITE  := http://www.gnu.org/software/automake
$(PKG)_OWNER    := https://github.com/starius
```
may also be useful, but the REAMDE is an excellent approach.

File: plugins/luarocks/lpeg.mk [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+$(PKG)_SUBDIR   := lpeg-$($(PKG)_VERSION)
+$(PKG)_FILE     := $($(PKG)_SUBDIR).tar.gz
+$(PKG)_URL      := http://www.inf.puc-rio.br/~roberto/$(PKG)/$($(PKG)_FILE)
+$(PKG)_DEPS     := gcc luarocks
+
+define $(PKG)_UPDATE
+    $(WGET) -q -O- 'http://www.inf.puc-rio.br/~roberto/lpeg/' | \
+    $(SED) -n 's,.*lpeg-\([0-9][^>]*\)\.tar.*,\1,p' | \
+    sort -h | tail -1
+endef
```
Comment (tonytheodore) [inline plugins/luarocks/lpeg.mk]: $(SORT)

File: plugins/luarocks/luarocks.mk [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+define $(PKG)_UPDATE
+    $(WGET) -q -O- 'https://keplerproject.github.io/luarocks/releases/' | \
+    $(SED) -n 's,.*luarocks-\([0-9][^>]*\)\.tar.*,\1,p' | \
+    sort -h | tail -1
+endef
```
Comment (tonytheodore) [inline plugins/luarocks/luarocks.mk]: $(SORT)

File: plugins/luarocks/llthreads2.mk [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+$(PKG)_CHECKSUM := 8c6fc7966cdcc15ae2f89f66ae72f6727a985e7d254f139ecf75a50956a3e8e4
+$(PKG)_SUBDIR   := lua-$(PKG)-$($(PKG)_VERSION)
+$(PKG)_FILE     := $($(PKG)_SUBDIR).tar.gz
+$(PKG)_URL      := https://github.com/moteus/lua-llthreads2/archive/v$($(PKG)_VERSION).tar.gz
+$(PKG)_DEPS     := gcc luarocks
+
+define $(PKG)_UPDATE
+    $(call MXE_GET_GITHUB_TAGS, moteus/lua-llthreads2) | \
+    $(SED) 's,^v,,g'
+endef
```
Comment (tonytheodore) [inline plugins/luarocks/llthreads2.mk]: Switch to new argument for `MXE_GET_GITHUB_TAGS` (also `lua-rapidjson.mk`, luasocket.mk)

**Comment by tonytheodore** [comment]
> Plugins currently require some boilerplate:

See https://github.com/mxe/mxe/commit/2944ccde5de26b062547ad87fb6354908f7aaf47, `$(PKG)_MAKEFILE` is no longer needed.

> How to build all packages from a particular plugin? from https://github.com/mxe/mxe/pull/935#issuecomment-160395601

There's also now an automatic `all-<plugin>` target which is the last segment of the plugin path. `all-rocks.mk` can be removed and `make all-luarocks MXE_PLUGIN_DIRS=plugins/luarocks ...` will work.

**Comment by starius** [comment]
I have taken into account all the comments.

Build on Wheezy produces broken lua.exe / lua53.dll. It can't find `libgcc_s_sjlj-1.dll`.

```
$ make all-luarocks MXE_PLUGIN_DIRS=plugins/luarocks MXE_TARGETS='i686-w64-mingw32.shared x86_64-w64-mingw32.shared'

$ ./usr/bin/i686-w64-mingw32.shared-lua 
err:module:import_dll Library libgcc_s_sjlj-1.dll (which is needed by L"Z:\\home\\mxe\\mxe-luarocks-review\\usr\\i686-w64-mingw32.shared\\bin\\lua53.dll") not found
err:module:import_dll Library lua53.dll (which is needed by L"Z:\\home\\mxe\\mxe-luarocks-review\\usr\\i686-w64-mingw32.shared\\bin\\lua.exe") not found
err:module:LdrInitializeThunk Main exe initialization for L"Z:\\home\\mxe\\mxe-luarocks-review\\usr\\i686-w64-mingw32.shared\\bin\\lua.exe" failed, status c0000135
```

**Comment by tonytheodore** [comment]
> Build on Wheezy produces broken lua.exe / lua53.dll. It can't find libgcc_s_sjlj-1.dll

Is it only Wheezy and/or lua.exe? The gcc *.dlls are [in a version-specific subdir](https://github.com/LuaAndC/mxe/commit/9020e9e0e53cd249a96fc48ee801dce6c28c9316) in case we were going to do side-by-side gcc versions (which now seems unlikely).

Does moving the *.dlls the the parent dir work?

**Comment by starius** [comment]
It doesn't depend on Wheezy/Jessie.

Clean installation:
```
$ make all-luarocks MXE_PLUGIN_DIRS=plugins/luarocks MXE_TARGETS='i686-w64-mingw32.shared x86_64-w64-mingw32.shared'
$ find usr/ -name libgcc_s_sjlj-1.dll
usr/i686-w64-mingw32.shared/bin/gcc-4.9.3/libgcc_s_sjlj-1.dll
```

Old installation (gcc rebuilt):
```
$ find usr/ -name libgcc_s_sjlj-1.dll
usr/i686-w64-mingw32.shared/bin/libgcc_s_sjlj-1.dll
usr/i686-w64-mingw32.shared/bin/gcc-4.9.3/libgcc_s_sjlj-1.dll
```

**Comment by tonytheodore** [comment]
I'm not sure what just happened. I tried creating a pull request using the web interface and it [created the branch](https://github.com/mxe/mxe/commits/gcc-dll) but seems to have merged it without a pull request https://github.com/mxe/mxe/commit/ae8f6def5cd64988c8c25e48bb2d4698dab45d8a.

**Comment by tonytheodore** [comment]
Reverted now https://github.com/mxe/mxe/commit/38a9ead6238a2155c10b6a991729b23360baae42, I'll do a proper fix tomorrow.

**Comment by starius** [comment]
For the record, the problem seems to be introduced by 9020e9e0e53cd249a96fc48ee801dce6c28c9316, #965.

**Comment by tonytheodore** [comment]
See #1032

**Comment by tonytheodore** [comment]
llthreads2, lpeg, and lua-rapidjson depend on luasocket:
```
$ make all-luarocks MXE_PLUGIN_DIRS=plugins/luarocks -j4
[build]     lua                    i686-w64-mingw32.shared
[done]      lua                    i686-w64-mingw32.shared                                2380 KiB       0m6.850s
[build]     luarocks               i686-w64-mingw32.shared
[done]      luarocks               i686-w64-mingw32.shared                                732 KiB        0m6.548s
[build]     llthreads2             i686-w64-mingw32.shared
[build]     lpeg                   i686-w64-mingw32.shared
[build]     lua-rapidjson          i686-w64-mingw32.shared
[build]     luasocket              i686-w64-mingw32.shared

Failed to build package llthreads2 for target i686-w64-mingw32.shared!
------------------------------------------------------------
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.shared-gcc -shared -o llthreads2.dll -L/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/lib src/l52util.obj src/llthread.obj -lkernel32 -llua
No existing manifest. Attempting to rebuild...
Updating manifest for /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/lib/luarocks/rocks
Error: rock_manifest file not found for luasocket scm-0 - not a LuaRocks 2 tree?
make[1]: *** [build-only-llthreads2_i686-w64-mingw32.shared] Error 1
real	0m6.847s
user	0m6.154s
sys	0m0.606s
------------------------------------------------------------
[log]      /Users/tonyt/dev/mxe/log/llthreads2_i686-w64-mingw32.shared

make: *** [/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/installed/llthreads2] Error 1
make: *** Waiting for unfinished jobs....

Failed to build package lpeg for target i686-w64-mingw32.shared!
------------------------------------------------------------
/Users/tonyt/dev/mxe/usr/bin/i686-w64-mingw32.shared-gcc -shared -o lpeg.dll -L/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/lib lpcap.obj lpcode.obj lpprint.obj lptree.obj lpvm.obj -llua
No existing manifest. Attempting to rebuild...
Updating manifest for /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/lib/luarocks/rocks
Error: rock_manifest file not found for luasocket scm-0 - not a LuaRocks 2 tree?
make[1]: *** [build-only-lpeg_i686-w64-mingw32.shared] Error 1
real	0m7.079s
user	0m6.417s
sys	0m0.575s
------------------------------------------------------------
[log]      /Users/tonyt/dev/mxe/log/lpeg_i686-w64-mingw32.shared

make: *** [/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/installed/lpeg] Error 1

Failed to build package lua-rapidjson for target i686-w64-mingw32.shared!
------------------------------------------------------------
-- Installing: /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/lib/luarocks/rocks/rapidjson/0.2.2-1/lib/rapidjson.dll
No existing manifest. Attempting to rebuild...
Updating manifest for /Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/lib/luarocks/rocks
Error: rock_manifest file not found for luasocket scm-0 - not a LuaRocks 2 tree?
make[1]: *** [build-only-lua-rapidjson_i686-w64-mingw32.shared] Error 1
real	0m11.512s
user	0m10.398s
sys	0m0.916s
------------------------------------------------------------
[log]      /Users/tonyt/dev/mxe/log/lua-rapidjson_i686-w64-mingw32.shared

make: *** [/Users/tonyt/dev/mxe/usr/i686-w64-mingw32.shared/installed/lua-rapidjson] Error 1
[done]      luasocket              i686-w64-mingw32.shared                                2340 KiB       0m12.423s
```

Easy fix https://github.com/tonytheodore/mxe/commit/297f697acac04859fd1387e5f8bf6d556b8421d1

**Comment by starius** [comment]
llthreads2, lpeg, and lua-rapidjson do not depend on luasocket. None of them needs network. Can you show logs for llthreads2, lpeg, and lua-rapidjson, please? Probably the build was failed because of race condition (`-j4`). Luarocks updates information about rocks installed and it seems to do it without synchronization between multiple processes. If it is true, some kind of mutex should be introduced to prevent this.

Another possible reason of printing `luasocket` in error logs of other rocks is that the build from the local rockspec failed for some reason and luarocks tried to download rocks from luarocks.org. Luarocks uses wget or luasocket for downloading. This behaviour should be suppressed in MXE.

**Comment by tonytheodore** [comment]
> Can you show logs for llthreads2, lpeg, and lua-rapidjson, please?

Only llthreads2 and lpeg this time: https://gist.github.com/tonytheodore/7765744f895cdad87035

> Luarocks uses wget or luasocket for downloading. This behaviour should be suppressed in MXE.

The [`PRELOAD` trick](https://github.com/mxe/mxe/commit/70c4d91a07046a2117feb4a05b7665c39bbd1965) is working on OSX now, disabling it always succeeds:
```
make all-luarocks MXE_PLUGIN_DIRS=plugins/luarocks -j4 PRELOAD=
```

**Comment by tonytheodore** [comment]
> Luarocks updates information about rocks installed and it seems to do it without synchronization between multiple processes. If it is true, some kind of mutex should be introduced to prevent this.

A workaround is to put a [`.NOTPARALLEL` target](http://www.gnu.org/software/make/manual/make.html#index-_002eNOTPARALLEL) in one of the files:
```makefile
diff --git a/plugins/luarocks/luarocks.mk b/plugins/luarocks/luarocks.mk
index b7cbbae..51209bb 100644
--- a/plugins/luarocks/luarocks.mk
+++ b/plugins/luarocks/luarocks.mk
@@ -18,6 +18,10 @@ define $(PKG)_UPDATE
     $(SORT) -h | tail -1
 endef

+# Luarocks updates information about rocks installed and it seems
+# to do it without synchronization between multiple processes.
+.NOTPARALLEL:
+
 # shared-only because Lua loads modules in runtime

 define $(PKG)_BUILD_SHARED
```

**Comment by starius** [comment]
I wrapped luarocks main tool with bash script which provides a mutex based on `mkdir`.
See http://wiki.bash-hackers.org/howto/mutex for the explanation
Please test if it works for you.

**Comment by tonytheodore** [comment]
Thanks, the lock dir needs to exist:
```diff
diff --git a/plugins/luarocks/luarocks.mk b/plugins/luarocks/luarocks.mk
index 10131ee..a852c37 100644
--- a/plugins/luarocks/luarocks.mk
+++ b/plugins/luarocks/luarocks.mk
@@ -39,6 +39,7 @@ define $(PKG)_BUILD_SHARED

     # create bash wrapper adding inter-process mutex
     # see https://github.com/mxe/mxe/pull/1017#issuecomment-161557440
+    $(INSTALL) -d '$(PREFIX)/$(TARGET)/lib/luarocks'
     (echo '#!/usr/bin/env bash'; \
      echo 'echo "== Using MXE wrapper: $(PREFIX)/$(TARGET)/bin/luarocks"'; \
      echo '# Creating a directory is an atomic operation, that is why'; \
```
and it all works well!

**Comment by starius** [comment]
Fixed.

**Comment by tonytheodore** [comment]
Do you want to do any rebasing? Otherwise I think this is ready to merge.

**Comment by starius** [comment]
Rebased

**Comment by tonytheodore** [comment]
Well done!

**Pull request merged**